pkg/process/metadata/parser/java/util.go

```diff
@@ -0,0 +1,126 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package javaparser
+
+import (
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/rickar/props"
+	"github.com/vibrantbyte/go-antpath/antpath"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type mapSource struct {
+	m map[string]string
+}
+
+// Get the value for the supplied key
+func (y *mapSource) Get(key string) (string, bool) {
+	val, ok := y.m[key]
+	return val, ok
+}
+
+// GetDefault gets the value for the supplied key or the defVal if missing
+func (y *mapSource) GetDefault(key string, defVal string) string {
+	val, ok := y.m[key]
+	if !ok {
+		return defVal
+	}
+	return val
+}
+
+func newArgumentSource(arguments []string, prefix string) *mapSource {
+	parsed := make(map[string]string)
+	for _, val := range arguments {
+		if strings.HasPrefix(val, prefix) {
+			parts := strings.SplitN(val[len(prefix):], "=", 2)
+			if len(parts) == 1 {
+				parsed[parts[0]] = ""
+			} else {
+				parsed[parts[0]] = parts[1]
+			}
+			continue
+		}
+	}
+	return &mapSource{parsed}
+}
+
+func newPropertySourceFromStream(rc io.ReadCloser, filename string) (*props.PropertyGetter, error) {
+	var properties props.PropertyGetter
+	var err error
+	if strings.HasSuffix(filename, ".properties") {
+		properties, err = props.Read(rc)
+	} else if strings.HasSuffix(filename, ".yaml") || strings.HasSuffix(filename, ".yml") {
+		properties, err = newYamlSource(rc)
+	} else {
+		err = fmt.Errorf("unhandled file type for %s", filename)
+	}
+	return &properties, err
+}
+
+func longestPathPrefix(pattern string) string {
+	idx := strings.IndexAny(pattern, "?*")
+	if idx < 0 {
+		return pattern
+	}
+	idx = strings.LastIndex(pattern[:idx], "/")
+	if idx < 0 {
+		return ""
+	}
+	return pattern[:idx]
+}
+
+func scanSourcesFromFileSystem(profilePatterns map[string][]string) map[string]*props.Combined {
```

Comment:
this code is very complex and nested. you have a for inside a for, heavily nested if clauses.
I think this code needs another pass. Split to sub-functions if needed, or try to reduce the complexity of the code

Reply:
@amarziali bump

Reply:
The algorithm is anyway complex itself because for every profile (that will generate different files) I have to take each pattern (that can match multiple files) and given this, since we're handling pattern, I extract the longest prefix not being a pattern starting walking from there to minimize fs scan. 
Other thing is that, when resolving properties, order matters hence I add in that precise order to the combined resolver (it's a chain resolver looking the next property source if the current is not containing a key)

Honestly, I applied some suggestions to improve the readability (i.e. move away inner methods) but for the complexity I really do not know if we can do better. I'm open to suggestions or we can conclude to keep as it otherwise